### DISCUSSION TITLE
"Possibly unbound" in else clause of while/else that must be bound

### DISCUSSION THREAD

**Discussion by dan-ri** [microsoft/pyright#4048] [Q&A] [answered]
Python 3.8.3, Pyright 1.1.275. This code is not the most readable, but it can't reference an unbound item, whereas pyright says `last_error` is possibly unbound:

```python
def do_something() -> None:
    pass


n = 0
while n < 5:
    try:
        do_something()
        break
    except Exception as error:
        n += 1
        last_error = error
else:
    raise last_error
```
Is this just a current inference limitation?

**Comment by erictraut** [comment, marked as answer]
This is expected. A static type checker like pyright cannot determine the number of times that a while loop is executed. It needs to assume that it might be executed zero times, in which case the `else` clause will be executed and the `last_error` symbol will be unbound. You'll need to modify your code to work within this limitation if you want to avoid the error.

**Reply by dan-ri** [reply]
Right, just thought some inference might be possible given the constant initializer and bound. Thanks @erictraut!
